### ISSUE TITLE
vmwarefusion driver hard coded to only look for docker-machine instances running on NAT vmnet interface

### ISSUE THREAD

**Issue by binarytemple** [docker/machine#2787]
I've been having a hard time figuring out why when I changed the networking type to vmnet2 (custom network configuration) docker-machine was unable to communicate with the machines. 

This got me interested in how docker-machine resolves the host IP under normal circumstances when the IP addresses change (as they do with every reboot).

Digging through the code, I came upon the following [code](https://github.com/docker/machine/blob/master/drivers/vmwarefusion/fusion_darwin.go#L464):

	// DHCP lease table for NAT vmnet interface
	var dhcpfile = "/var/db/vmware/vmnet-dhcpd-vmnet8.leases"

	if vmxfh, err = os.Open(d.vmxPath()); err != nil {
		return "", err
	}

So that's how the driver resolves the changed IP address, by parsing that file. The problem of course is that the new dhcp lease will reside in the file "/var/db/vmware/vmnet-dhcpd-vmnet2.leases" hence docker-machine will be unable to determine the new IP address. 

I'd provide a patch, but unfortunately at this stage my golang programming skills are absolutely non-existent.

**Comment by binarytemple** [comment]
Yeah nice work @jeanlaurent / @daehyeok .. I just took the changes built and tested locally. It works correctly now. No problem connecting to the machines. I needed to regenerate the server certificates but that was expected.

**Comment by frapposelli** [comment]
@binarytemple can you describe your use case for changing the network config?

**Comment by binarytemple** [comment]
Sure. As far as I can tell, there is no way to specify a default network configuration when creating new machines using `vmrun` (which is what docker-machine uses).

I would like to create the machines using a custom network connection so that the (docker-machine, and other) virtual machines are visible to one another, with a gateway providing NAT. In this case the custom network is called vmnet2. 

But when the machines are created, they are created using the default network (I can't find a way to change this) called 'Share with my Mac' or vmnet8.

After creating the machine, I then need to go to it's preferences and change the settings appropriately. Of course this still causes problems with swarm etc, but I'm looking into them now.

TLDR;

Lame VMware network configuration settings on OSX.

**Comment by frapposelli** [comment]
The default `vmnet8` does exactly that, it's a network segment that is shared between all the VMs attached to it and is NAT'd on the outside by your Mac that act as a gateway. Did that not work for you?

Custom networks are usually created to segregate network segments (e.g. non routed private segments) or where you want to provide your own NAT (e.g. on a separate VM).

**Comment by binarytemple** [comment]
I will need to rerun my tests, but if I recall correctly I wasn't able to connect to exposed ports on boot2docker VMS despite enabling NAT in the VMware configuration files.(IMHO fusion is a dog, I'm only paying for it rather than using VirtualBox because I'll be deploying to a large VSphere cluster and want to get a feel for the CLI/tooling sooner than later).So I opted for the custom configuration with an ip address per VM (in a range of my choice) and NAT through the OSX host. And that works with the patch provided. I can access exposed ports via the boot2docker ip addresses (running multiple instances) and port rather than talking to NAT'd ports on local adaptors or resorting to SSH tunnels. There's still some clunky manual work to do because you can't specify network at VM creation time (VMwares fault), but it seems to be largely working now (will explain process if required).I also wanted the machines on a specific sub-net 10.69.1.0 which is the same or similar to what I'll run in production.I'd guess my plan would be to explore docker overlay network ‎later but for now that's where I'm at.I'm on VMware fusion pro 8.1.0 BTW.Does that sound OK? Bryan                                                                                                                                                                   From: Fabio RapposelliSent: Monday, January 11, 2016 10:51 PMTo: docker/machine‎Reply To: docker/machineCc: Bryan HuntSubject: Re: [machine] vmwarefusion driver hard coded to only look for docker-machine instances running on NAT vmnet interface (#2787)‎The default vmnet8 does exactly that, it's a network segment that is shared between all the VMs attached to it and is NAT'd on the outside by your Mac that act as a gateway. Did that not work for you?

Custom networks are usually created to segregate network segments (e.g. non routed private segments) or where you want to provide your own NAT (e.g. on a separate VM).

**Comment by frapposelli** [comment]
First, I'm not against making the change in the code (I gave LGTM to #2790) but I want to better understand the rationale behind it, and if there are any use cases that are not covered by the current implementation.

From your message your requirements seem to be:
- The boot2docker VMs must be connected on the same subnet and reachable (from your Mac and between them)
- The VMs must be on a specific subnet (10.69.1.0/?)
- You want to do port address translation (e.g. port 4000 on your mac -> port 80 on a b2d VM)

Is that correct?

**Comment by binarytemple** [comment]
The boot2docker VMs must be connected on the same subnet and reachable (from your Mac and between them)The VMs must be on a specific subnet (10.69.1.0/?)You want to do port address translation (e.g. port 4000 on your mac -> port 80 on a b2d VM)Correct for the first two - ‎but if I have them working there is no need for 3 (port address translation/DNAT) (it might be nice in some cases but not necessary for my development environment). Thanks, Bryan                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         From: Fabio RapposelliSent: Tuesday, January 12, 2016 9:18 AMTo: docker/machineReply To: docker/machineCc: Bryan HuntSubject: Re: [machine] vmwarefusion driver hard coded to only look for docker-machine instances running on NAT vmnet interface (#2787)First, I'm not against making the change in the code (I gave LGTM to #2790) but I want to better understand the rationale behind it, and if there are any use cases that are not covered by the current implementation.

From your message your requirements seem to be:


The boot2docker VMs must be connected on the same subnet and reachable (from your Mac and between them)
The VMs must be on a specific subnet (10.69.1.0/?)
You want to do port address translation (e.g. port 4000 on your mac -> port 80 on a b2d VM)


Is that correct?

—Reply to this email directly or view it on GitHub.

**Comment by frapposelli** [comment]
Got it, thanks.

You can definitely do 1 and 2 without creating a custom network:

- VMs connected on `vmnet8` are not restricted and can talk to each other

```shell
~ ⟩ docker-machine ls
NAME       ACTIVE   DRIVER         STATE     URL                        SWARM
buildbox   -        vmwarefusion   Stopped
test1      -        vmwarefusion   Running   tcp://172.16.94.140:2376
test2      -        vmwarefusion   Running   tcp://172.16.94.141:2376
test3      -        vmwarefusion   Running   tcp://172.16.94.142:2376
~ ⟩ docker-machine ssh test1 ping -c 3 172.16.94.141
PING 172.16.94.141 (172.16.94.141): 56 data bytes
64 bytes from 172.16.94.141: seq=0 ttl=64 time=0.240 ms
64 bytes from 172.16.94.141: seq=1 ttl=64 time=0.238 ms
64 bytes from 172.16.94.141: seq=2 ttl=64 time=0.294 ms

--- 172.16.94.141 ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 0.238/0.257/0.294 ms
~ ⟩ docker-machine ssh test1 ping -c 3 172.16.94.142
PING 172.16.94.142 (172.16.94.142): 56 data bytes
64 bytes from 172.16.94.142: seq=0 ttl=64 time=0.238 ms
64 bytes from 172.16.94.142: seq=1 ttl=64 time=0.303 ms
64 bytes from 172.16.94.142: seq=2 ttl=64 time=0.313 ms

--- 172.16.94.142 ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 0.238/0.284/0.313 ms
```

- By default when you install Fusion `vmnet1` and `vmnet8` are configured with a range that should not collide with any currently attached network, but the subnet is easily configurable, you can move your `vmnet8` to `10.69.1.0` by following this KB: http://kb.vmware.com/kb/1026510

That said, I'm ok with going forward with the code changes and start adding support for custom networking options.

**Comment by jeanlaurent** [comment]
@frapposelli when you say 
> That said, I'm ok with going forward with the code changes and start adding support for custom networking options.

You're talking about #2790 right ?

**Comment by frapposelli** [comment]
@jeanlaurent yup :+1:
